### PR TITLE
Make all template strings configurable

### PR DESCRIPTION
**Pull request by Jaza**
Give template strings default values, and allow those
defaults to be overridden with these config options:

- MAILGUN_API_URL_TEMPLATE
- MAILGUN_DEBUG_TEMPLATE
- MAILGUN_LOGGING_TEMPLATE
- MAILGUN_ERROR_SUBJECT_WITHEXCINFO_TEMPLATE
- MAILGUN_ERROR_SUBJECT_WITHOUTEXCINFO_TEMPLATE
- MAILGUN_ERROR_TEMPLATE
- MAILGUN_API_ERROR_TEMPLATE

### CODE REVIEW

File: flask_mailgun.py [hunk 1]
```diff
@@ -49,21 +58,36 @@
...
+        self.api_error_template = None
+
         if app is not None:
             self.init_app(app)
 
     def init_app(self, app):
         """Inititialize the extension.
 
-        There are two configuration options for authentization:
+        There are two configuration options for authentication:
```
Comment (keosak) [inline flask_mailgun.py]: Good catch, I mixed up authentication with authorization in my head into one word. Perhaps this should be in a separate commit?

File: flask_mailgun.py [hunk 2]
```diff
@@ -73,18 +97,72 @@ def init_app(self, app):
...
+          - MAILGUN_ERROR_TEMPLATE
+          - MAILGUN_API_ERROR_TEMPLATE
         """
         app.extensions['mailgun'] = self
         self.debug = app.debug
+
+        self.domain = app.config['MAILGUN_DOMAIN']
+        self.key = app.config['MAILGUN_KEY']
+
+        self.api_url_template = app.config.get(
```
Comment (keosak) [inline flask_mailgun.py:116]: I think the more Pythonic way to do this is the following. This also applies to all instances of the pattern below. It makes the code shorter with fewer conditionals, and that's always nice.

```python
self.api_url_template = app.config.get('MAILGUN_API_URL_TEMPLATE', default_api_url_template)
```

File: flask_mailgun.py [hunk 3]
```diff
@@ -101,27 +179,33 @@ def send(self, **data):
...
         data['from'] = from_
+
         if 'html' in data and 'text' not in data:
             data['text'] = html2text(data['html'])
+
         if self.debug:
             data.setdefault('html', '(no HTML)')
             current_app.logger.debug(debug_template, data)
             return
-        url = 'https://api.mailgun.net/v3/{}/messages'.format(self.domain)
```
Comment (keosak) [inline flask_mailgun.py:111]: This is the official Mailgun API endpoint URL. Why would you need a different one? Different version perhaps? In that case, I think it would be more convenient to have just the version in configuration.

File: flask_mailgun.py [hunk 4]
```diff
@@ -101,27 +179,33 @@ def send(self, **data):
...
+
+        url = self.api_url_template.format(domain=self.domain)
+
         response = requests.post(url, auth=('api', self.key), data=data)
         if response.status_code != 200:
-            raise APIError(response)
-        current_app.logger.info('Mailgun sent to %s', data['to'])
+            raise APIError(response, self.api_error_template)
 
+        current_app.logger.info(
```
Comment (keosak) [inline flask_mailgun.py]: The logging module accepts the formatting string separate from the arguments for a reason. It formats the string only if it will be actually logged. I know it looks weird when the whole rest of the code uses different formatting method, but that's just how it is.

**Comment by keosak** [comment]
Hello @Jaza. I wrote this code to be used in my company. I was pleasantly surprised that someone else would find it useful. I won't merge the PR but I would like to give you some constructive feedback as to why. I made some comments on the commit, but those are minor issues. My main objections are in the following paragraphs. You are of course welcome to just use your own fork, the code is open source.

You have added seven instance attributes to the `Mailgun` class and seven configuration options. That is too much for my taste. Also, these templates have several different formats and that makes it confusing. I would rather suggest the following.

Move `debug_template`, `logging_template` and `error_template` to be class instances of the `Mailgun` class. To override them, just make a subclass.

```python
class Mailgun:

    debug_template = """\
...
"""

    logging_template = """\
...
"""

    error_template = """\
...
"""
```

```python
from flask_mailgun import Mailgun as BaseMailgun

class Mailgun(BaseMailgun):

    debug_template = """\
MY DEBUG TEMPLATE
"""
```

Don't bother with the `APIError` message. Buit-in Python exceptions are not customizable and I don't think I have seen a popular Python library offer this option either. The same goes for subject templates.

I commented on the API endpoint template on the commit. Is it necessary to have the whole URL there? Maybe it is. That is the only configuration option I would add to the `app.config` dictionary.

**Comment by keosak** [comment]
Also, for the API endpoint URL configuration option I would use named parameter, eg. `https://api.mailgun.net/v3/{domain}/messages`.

File: flask_mailgun.py (hunk 2 above)
Comment (Jaza) [reply flask_mailgun.py:116]: Ok, I have changed this as you requested. I realise that the one-line version is more Pythonic, but the "is None" version lets people override these to be an empty string if they want (although not sure why anyone would want that).

File: flask_mailgun.py (hunk 3 above)
Comment (Jaza) [reply flask_mailgun.py:111]: Yeah, could be many reasons. Mailgun changing the URL on their end suddenly? Overriding / aliasing the Mailgun URL using local DNS in your environment? Setting up a "fake Mailgun" for test purposes in your environment? But really, the main reason (in my opinion) is that a URL should always be configurable somehow, simply as a matter of principle, it shouldn't be hard-coded into a library like this with no way of overriding it.

**Comment by Jaza** [comment]
Hi @keosak , thanks for writing this library, I'm finding it useful for a Flask site that I'm working on, where email sending via Mailgun is required.

Per your feedback, I have moved the various template strings to be class attributes, that can be overridden via subclassing, instead of being Flask config variables. Personally I prefer config variables for things like this, but I'm happy to subclass instead, my main concern is that these strings can be overridden one way or another. This reduces the size of the change significantly.

I have removed the configurable `APIError` message.

Actually, the main thing that I require be configurable, is `error_subject_withexcinfo_template` and `error_subject_withexcinfo_template`. I always like to put the site name in the subject line, because I maintain numerous sites, and I like to see in the email subject which site I'm receiving an error report for.

I hope you're willing to merge my PR after having made these changes. If not, then I will have to use my fork instead of your version.

**Comment by keosak** [comment]
The changes look OK and I will merge the PR. Thank you for your contribution. I have just one final note.

Aside from the semantic change, ie. making the templates configurable, you also made some cosmetic changes in other parts of code. I mean adding and removing blank lines, breaking long lines, etc. This project doesn't have a declared style guide or automated code formatter settings, but many other projects do. Even when they are not there, however, you would do yourself well to keep your contributions consistent with the overall style of the code base. Consistency is key to readability, and keeping commits short by avoiding unnecessary cosmetic changes makes reviewing them easier.

Again, thank you and I hope your project is successful.
